kani-compiler/src/kani_middle/transform/check_uninit/ty_layout.rs

```diff
@@ -0,0 +1,328 @@
+                        .map(|mut req| {
+                            req.offset += elem_offset;
+                            req
+                        })
+                        .collect::<Vec<_>>();
+                    result.append(&mut next_data_bytes)
+                }
+            }
+            Ok(result)
+        }
+        FieldsShape::Arbitrary { ref offsets } => {
+            match ty.kind().rigid().expect(&format!("unexpected type: {ty:?}")) {
+                RigidTy::Adt(def, args) => {
+                    match def.kind() {
+                        AdtKind::Enum => {
+                            // Support basic enumeration forms
+                            let ty_variants = def.variants();
+                            match layout.variants {
+                                VariantsShape::Single { index } => {
+                                    // Only one variant is reachable. This behaves like a struct.
+                                    let fields = ty_variants[index.to_index()].fields();
+                                    let mut fields_data_bytes = vec![];
+                                    for idx in layout.fields.fields_by_offset_order() {
+                                        let field_offset = offsets[idx].bytes();
+                                        let field_ty = fields[idx].ty_with_args(&args);
+                                        fields_data_bytes.append(&mut data_bytes_for_ty(
+                                            machine_info,
+                                            field_ty,
+                                            field_offset + current_offset,
+                                        )?);
+                                    }
+                                    Ok(fields_data_bytes)
+                                }
+                                VariantsShape::Multiple {
+                                    tag_encoding: TagEncoding::Niche { .. },
+                                    ..
+                                } => {
+                                    Err(format!("Unsupported Enum `{}` check", def.trimmed_name()))?
+                                }
+                                VariantsShape::Multiple { variants, tag, .. } => {
+                                    // Retrieve data bytes for the tag.
+                                    let tag_size = match tag {
+                                        Scalar::Initialized { value, .. } => {
+                                            value.size(&machine_info)
+                                        }
+                                        Scalar::Union { .. } => {
+                                            unreachable!("Enum tag should not be a union.")
+                                        }
+                                    };
+                                    let tag_data_bytes =
+                                        vec![DataBytes { offset: current_offset, size: tag_size }];
```

Comment:
Is it safe to assume that the tag is always at offset 0?

Reply:
Not if the enum is a field in another struct, for example

Reply:
I mean, 0 for the current type (i.e. `current_offset`). Can you add an assertion here too?

Reply:
Added an assertion